### ISSUE TITLE
Crash

### ISSUE THREAD

**Issue by opk12** [streetcomplete/StreetComplete#5241]
It's a while I did not use StreetComplete. I opened it, it was at a very low zoom (showed Europe, Africa, part of America) and I zoomed a bit [around South Egypt](https://www.openstreetmap.org/#map=9/24.1242/32.8656). The app crashed.

```
java.lang.RuntimeException: Tried to perform an operation on an invalid pointer! This means you may have used a MapData that has already been removed.
at com.mapzen.tangram.MapData.checkPointer(MapData.java:116)
at com.mapzen.tangram.MapData.setFeatures(MapData.java:36)
at de.westnordost.streetcomplete.screens.main.map.components.DownloadedAreaMapComponent.set(DownloadedAreaMapComponent.kt:32)
at de.westnordost.streetcomplete.screens.main.map.DownloadedAreaManager$update$1.invokeSuspend(DownloadedAreaManager.kt:51)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@6afc11a, Dispatchers.Default]
```

**Versions:**
StreetComplete 53.3 from F-droid on Android 7.1

**Comment by opk12** [comment]
> Currently I'm trying to have the `set` `synchronized`...

Regardless of this issue: Multi-threaded access to shared resources must always be inside `synchronized` (or the field must be `volatile`, which is effectively a shorthand) because the JVM (this is not Kotlin-specific) can legally
* do things non-atomically (writing an object reference into a field in two steps)
* reorder instructions (next line is executed before previous line)
* have each thread work on its local copy of the heap and not see updates from other threads

From a quick google, Kotlin appears not to change this (seems just an unreadable syntax around the same JVM?) so if it's multithreaded, that should be synchronized right away.

As a general rule, it's `synchronized(this.some_private_object) {...}` not `synchronized(this) { ... }` just to make sure a client can't synchronize on the same object, messing with the lock.

These and other common mistakes are covered in the well-known book `Effective Java` by Joshua Bloch (I think that synchronization is Item 66).

**Comment by Helium314** [comment]
I managed to reproduce the crash by starting SC completely zoomed out with location off. Opening the undo history zooms to the position of the last edit and triggers the crash (a slow phone and a debug APK might help).

Using `synchronized` indeed seems to help (never crashed).

**Comment by Helium314** [comment]
> I thought that the problem may be related to some Android lifecycle changes where the entire tangram MapController is disposed of.

I assume that could still trigger the crash, but more rarely and (presumably) only while the app is being closed.

**Comment by westnordost** [comment]
Possibly fixed by #5269
